### PR TITLE
Check the socket state without race coniditons

### PR DESCRIPTION
**Pull request by lrh2000**
The code introduced in https://github.com/asterinas/asterinas/pull/1434 (cc @StevenJiang1110) contains a terrible race condition:

https://github.com/asterinas/asterinas/blob/e6c613f53841983765a7b3c56ea9958775c76199/kernel/src/net/socket/ip/stream/connected.rs#L77-L98

For example, it is possible that `recv` on a newly accepted socket returns `InvalidState` on line 78, then the socket is established between lines 78~92 (e.g. the IRQ occurs in the middle), then the `before_established` check on line 92 fails, so we return `ECONNRESET` to userspace.

Besides:
https://github.com/asterinas/asterinas/blob/e6c613f53841983765a7b3c56ea9958775c76199/kernel/src/net/socket/ip/stream/connected.rs#L195
 - How can `SynSent` be possible?
https://github.com/asterinas/asterinas/blob/e6c613f53841983765a7b3c56ea9958775c76199/kernel/src/net/socket/ip/stream/connected.rs#L198
 - Why not do this? This generally makes our code simpler, I believe.

### CODE REVIEW

**Comment by StevenJiang1110** [comment]
Thanks for your fix! It's my fault to introduce such a race condition.

> How can SynSent be possible?

Yes, only `SynReceived` is possible. 

> Why not do this? This generally makes our code simpler, I believe.

I didn't check the behavior in Linux at the time, and I failed to recognize the race condition. 

However, I believe it would be more appropriate to check the established state in accept after doing some search, as mentioned in [this blog](https://www.alibabacloud.com/blog/tcp-syn-queue-and-accept-queue-overflow-explained_599203).

**Comment by lrh2000** [comment]
> However, I believe it would be more appropriate to check the established state in `accept` after doing some search, as mentioned in [this blog](https://www.alibabacloud.com/blog/tcp-syn-queue-and-accept-queue-overflow-explained_599203).

Agree. This seems to be the scenario, at least when the TCP Fast Open mechanism is off. I checked several online sources and they claim the same thing.

I also tried to check the [Linux kernel source code](https://elixir.bootlin.com/linux/v6.11.8/source/net/ipv4/tcp_input.c#L7304). However, the Linux implementation is too complex, with many confusing method names, making it hard to follow. I didn't get a clue to understand it, at least for now.

**Review by StevenJiang1110 — APPROVED** [review]
LGTM!
